locator/token_metadata.cc

```diff
@@ -763,122 +713,108 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
-        calculate_pending_ranges_for_leaving(unpimplified_this, strategy, new_pending_ranges, all_left_metadata);
-        // At this stage newPendingRanges has been updated according to leave operations. We can
-        // now continue the calculation by checking bootstrapping nodes.
-        calculate_pending_ranges_for_bootstrap(strategy, new_pending_ranges, all_left_metadata, get_dc_rack);
-        all_left_metadata->clear_gently().get();
-
-        // At this stage newPendingRanges has been updated according to leaving and bootstrapping nodes.
-        set_pending_ranges(keyspace_name, std::move(new_pending_ranges), can_yield::yes);
+        // new_token_metadata incorporates all the changes from leaving, bootstrapping and replacing
+        const auto new_token_metadata = token_metadata([&]() -> std::unique_ptr<token_metadata_impl> {
+            auto result = clone_only_token_map(false).get0();
+
+            // construct new_normal_tokens based on _bootstrap_tokens and _replacing_endpoints
+            std::unordered_map<inet_address, std::unordered_set<token>> new_normal_tokens;
+            if (!_replacing_endpoints.empty()) {
+                for (const auto& [token, inet_address]: _token_to_endpoint_map) {
+                    const auto it = _replacing_endpoints.find(inet_address);
+                    if (it == _replacing_endpoints.end()) {
+                        continue;
+                    }
+                    new_normal_tokens[it->second].insert(token);
+                }
+                for (const auto& [replace_from, replace_to]: _replacing_endpoints) {
+                    result->remove_endpoint(replace_from);
+                }
+            }
+            for (const auto& [token, inet_address]: _bootstrap_tokens) {
+                new_normal_tokens[inet_address].insert(token);
+            }
+            // apply new_normal_tokens
+            for (auto& [endpoint, tokens]: new_normal_tokens) {
+                result->update_topology(endpoint,
+                    get_dc_rack(endpoint),
+                    _normal_token_owners.contains(endpoint) ? node::state::normal : node::state::joining);
+                result->update_normal_tokens(std::move(tokens), endpoint).get();
+            }
+            // apply leaving endpoints
+            for (const auto& endpoint: _leaving_endpoints) {
+                result->remove_endpoint(endpoint);
+            }
+            result->sort_tokens();
+            return result;
+        }());
+
+        // merge tokens from token_to_endpoint and bootstrap_tokens,
+        // preserving tokens of leaving endpoints
+        const auto tokens = [&]() -> std::vector<dht::token> {
+            auto result = std::vector<dht::token>();
+            result.reserve(get_token_to_endpoint().size() + get_bootstrap_tokens().size());
+            for (const auto& p: get_token_to_endpoint()) {
+                result.push_back(p.first);
+            }
```

Comment:
Why not simply copy `sorted_tokens()`?

Reply:
Is it simpler? we still need bootstrap tokens, similar looking loops read easier to me.

Reply:
sorted_token can be copied in bulk using std::copy soit's both simpler and more efficient.

Reply:
ok, will update

Reply:
If we care about efficiency like this, the following algorithm would be appropriate:
0. reserve space in `result` vector for all tokens
1. copy `sorted_tokens()` to the first part of `result` vector
2. copy `bootstrap_tokens` to the second part of `result` vector
3. sort only the second part
4. `std::merge` the two parts

Reply:
> If we care about efficiency like this

I definitely don't, the complexity/utility ratio is just too high.